app/controllers/api/v1/pushes_controller.rb

```diff
@@ -0,0 +1,218 @@
+# frozen_string_literal: true
+
+require "securerandom"
+
+class Api::V1::PushesController < Api::BaseController
+  include SetPushAttributes
+  include LogEvents
+
+  helper UrlsHelper
+
+  before_action :set_current_kind
+  before_action :set_push, only: %i[show preview audit destroy]
+
+  def show
+    # This push may have expired since the last view.  Validate the url
+    # expiration before doing anything.
+    @push.check_limits
+
+    if @push.expired
+      log_view(@push)
+      render json: @push.to_json(payload: true)
+      return
+    end
+
+    if @push.passphrase.present?
+      # JSON requests must pass the passphrase in the params
+      has_passphrase = params[:passphrase] == @push.passphrase
+
+      unless has_passphrase
+        # Passphrase hasn't been provided or is incorrect
+        # Passphrase hasn't been provided or is incorrect
+        render json: {error: "This push has a passphrase that was incorrect or not provided."}
+        return
+      end
+    end
+
+    log_view(@push)
+    expires_now
+
+    render json: @push.to_json(payload: true)
+
+    # If files are attached, we can't expire immediately as the viewer still needs
+    # to download the files.  In the case of files, this push will be expired on the
+    # next ExpirePushesJob run or next view attempt.  Whichever comes first.
+    if !@push.files.attached? && !@push.views_remaining.positive?
+      # Expire if this is the last view for this push
+      @push.expire!
+    end
+  end
+
+  def create
+    # Require authentication if allow_anonymous is false
+    # See config/settings.yml
+    authenticate_user! if Settings.enable_logins && !Settings.allow_anonymous
+
+    @push = Push.new(push_params)
+    @push.kind = @current_kind
+
+    @push.user_id = current_user.id if user_signed_in?
+
+    create_detect_deletable_by_viewer(@push, push_params)
+    create_detect_retrieval_step(@push, push_params)
+
+    if @push.save
+      log_creation(@push)
+
+      render json: @push, status: :created
+    else
+      render json: @push.errors, status: :unprocessable_entity
+    end
+  end
+
+  def preview
+    @secret_url = helpers.secret_url(@push)
+    render json: {url: @secret_url}, status: :ok
+  end
+
+  def audit
```

Comment:
The response of this action should be same as before. So, a bit complex logic is added for json_logs.